amethyst_renderer_new/src/graph/build.rs

```diff
@@ -0,0 +1,575 @@
+                    layouts: if colors[i].clear.is_some() {
+                        image::ImageLayout::Undefined
+                    } else {
+                        image::ImageLayout::General
+                    }..image::ImageLayout::General,
+                }
+            });
+
+            // Configure depth-stencil attachments last
+            let depth_stencil = pass.depth_stencil().clone().map(|format| {
+                let depth_stencil = depth_stencil.as_ref().unwrap();
+                assert_eq!(depth_stencil.format, format);
+                pass::Attachment {
+                    format,
+                    ops: pass::AttachmentOps {
+                        load: if depth_stencil.clear.is_some() {
+                            pass::AttachmentLoadOp::Clear
+                        } else {
+                            pass::AttachmentLoadOp::Load
+                        },
+                        store: pass::AttachmentStoreOp::Store,
+                    },
+                    stencil_ops: pass::AttachmentOps::DONT_CARE,
+                    layouts: if depth_stencil.clear.is_some() {
+                        image::ImageLayout::Undefined
+                    } else {
+                        image::ImageLayout::General
+                    }..image::ImageLayout::General,
+                }
+            });
+
+            let depth_stencil_ref = depth_stencil.as_ref().map(|_| {
+                (
+                    inputs.len() + colors.len(),
+                    image::ImageLayout::DepthStencilAttachmentOptimal,
+                )
+            });
+
+            // Configure the only `Subpass` using all attachments
+            let subpass = pass::SubpassDesc {
+                colors: &(0..colors.len())
+                    .map(|i| {
+                        (i + inputs.len(), image::ImageLayout::ColorAttachmentOptimal)
+                    })
+                    .collect::<Vec<_>>(),
+                depth_stencil: depth_stencil_ref.as_ref(),
+                inputs: &(0..inputs.len())
+                    .map(|i| (i, image::ImageLayout::ShaderReadOnlyOptimal))
+                    .collect::<Vec<_>>(),
+                preserves: &[],
+            };
+
+            // Create `RenderPass` with all attachments
+            // and single `Subpass`
+            device.create_render_pass(
+                &inputs
+                    .chain(colors)
+                    .chain(depth_stencil)
+                    .collect::<Vec<_>>(),
+                &[subpass],
+                &[], // TODO: Add external subpass dependency
+            )
+        };
+
+        // Create `DescriptorSetLayout` from bindings
+        let descriptor_set_layout = device.create_descriptor_set_layout(pass.bindings());
+
+        // Create `PipelineLayout` from single `DescriptorSetLayout`
+        let pipeline_layout = device.create_pipeline_layout(&[&descriptor_set_layout]);
```

Comment:
Pipeline layouts and pipelines are not necessarily strongly-coupled. A pipeline layout can be compatible with multiple pipelines. Switching (non-compatible) pipeline layouts introduces costs due to rebinding descriptor sets and push constants.

Reply:
But how often different pipelines can share pipeline layout?
Doesn't pipeline layout defines bindigs interface of shaders?
How often different shaders share same bindings interface?

Reply:
Yes, it defines the interface but can be a superset (some sort of 'uberlayout'). There is always a tradeoff between both approaches: minimal pipeline layout vs 'uberlayout'. Which to choose depends on the available pipelines. It might be interesting when having a lot of similar materials with lightly varying pipelines. I would say that expecially latest gen architectures can live with only a few pipeline layouts by using large descriptor sets, only bound once, but I'm not an expert.

Reply:
This makes sense but I'd like to hide such details from user. So I need to magically create super-layout for consecutive of passes.